File: nsynth/verify_stage4.py
#!/usr/bin/env python3
"""
Stage 4 End-to-End Synthesis Verification

Verifies time-parameterized benchmarks (fibonacci, factorial, triangular, linear, poly)
by extracting problem definitions, calling mog_synth, verifying output, and reporting
solve times and methods used.
"""

import json
import subprocess
import sys
import time
from pathlib import Path
from dataclasses import dataclass, asdict
from typing import Optional

@dataclass
class SynthesisResult:
    problem_name: str
    variant: int
    success: bool
    solve_time: float
    method: str
    code: Optional[str]
    error: Optional[str]

def get_benchmark_list() -> list[dict]:
    """Extract benchmark list from Rust benchmark module."""
    # Run the binary to extract benchmark metadata
    result = subprocess.run(
        ["./target/release/mog_synth", "--list-benchmarks"],
        capture_output=True,
        text=True,
        cwd="/Users/bobbyprice/projects/nCPU/nsynth"
    )

    if result.returncode != 0:
        # Fallback: manually define the representative benchmarks
        return [
            {"name": "fibonacci", "variant": 0},
            {"name": "factorial", "variant": 0},
            {"name": "triangular_number", "variant": 0},
            {"name": "linear_series", "variant": 0},
            {"name": "polynomial_eval", "variant": 0},
        ]

    try:
        return json.loads(result.stdout)
    except:
        return [
            {"name": "fibonacci", "variant": 0},
            {"name": "factorial", "variant": 0},
            {"name": "triangular_number", "variant": 0},
            {"name": "linear_series", "variant": 0},
            {"name": "polynomial_eval", "variant": 0},
        ]

def run_synthesis(problem_name: str, variant: int) -> SynthesisResult:
    """Run mog_synth on a single problem and return results."""
    problem_spec = {
        "name": problem_name,
        "variant": variant,
    }

    start = time.time()
    try:
        result = subprocess.run(
            ["./target/release/mog_synth",
             "--problem-name", problem_name,
             "--problem-variant", str(variant),
             "--json"],
            capture_output=True,
            text=True,
            timeout=300,
            cwd="/Users/bobbyprice/projects/nCPU/nsynth"
        )
        elapsed = time.time() - start

        if result.returncode != 0:
            return SynthesisResult(
                problem_name=problem_name,
                variant=variant,
                success=False,
                solve_time=elapsed,
                method="error",
                code=None,
                error=result.stderr or result.stdout
            )

        try:
            output = json.loads(result.stdout)
            return SynthesisResult(
                problem_name=problem_name,
                variant=variant,
                success=output.get("success", False),
                solve_time=output.get("solve_time", elapsed),
                method=output.get("method", "unknown"),
                code=output.get("code"),
                error=output.get("error")
            )
        except json.JSONDecodeError:
            return SynthesisResult(
                problem_name=problem_name,
                variant=variant,
                success=False,
                solve_time=elapsed,
                method="parse_error",
                code=None,
                error=f"Failed to parse JSON output: {result.stdout}"
            )

    except subprocess.TimeoutExpired:
        elapsed = time.time() - start
        return SynthesisResult(
            problem_name=problem_name,
            variant=variant,
            success=False,
            solve_time=elapsed,
            method="timeout",
            code=None,
            error="Synthesis timed out after 300 seconds"
        )
    except Exception as e:
        elapsed = time.time() - start
        return SynthesisResult(
            problem_name=problem_name,
            variant=variant,
            success=False,
            solve_time=elapsed,
            method="exception",
            code=None,
            error=str(e)
        )

def verify_compiled_code(code: str, problem_name: str) -> bool:
    """Try to compile and run the synthesized code."""
    if not code:
        return False

    # Create a temporary test file
    test_file = Path(f"/tmp/test_{problem_name}.mog")
    try:
        test_file.write_text(code)

        # Try to run it with mog_synth interpreter
        result = subprocess.run(
            ["./target/release/mog_synth", "--run-file", str(test_file)],
            capture_output=True,
            text=True,
            timeout=10,
            cwd="/Users/bobbyprice/projects/nCPU/nsynth"
        )

        return result.returncode == 0
    except Exception as e:
        print(f"  Verification failed: {e}", file=sys.stderr)
        return False
    finally:
        try:
            test_file.unlink()
        except:
            pass

def main():
    print("=" * 70)
    print("Stage 4 Time-Parameterized Synthesis Verification")
    print("=" * 70)

    # Representative benchmarks to verify
    benchmarks = [
        ("fibonacci", 0),
        ("factorial", 0),
        ("triangular_number", 0),
        ("linear_series", 0),
        ("polynomial_eval", 0),
    ]

    results: list[SynthesisResult] = []
    method_counts: dict[str, int] = {}

    for problem_name, variant in benchmarks:
        print(f"\nSynthesizing {problem_name} (variant {variant})...")
        result = run_synthesis(problem_name, variant)
        results.append(result)

        # Update method distribution
        method = result.method if result.success else "failed"
        method_counts[method] = method_counts.get(method, 0) + 1

        # Print result
        status = "✓ PASS" if result.success else "✗ FAIL"
        print(f"  {status} | Time: {result.solve_time:.2f}s | Method: {result.method}")

        if result.error:
            print(f"  Error: {result.error[:100]}")

        if result.code:
            print(f"  Code length: {len(result.code)} bytes")
            # Verify compilation
            if verify_compiled_code(result.code, problem_name):
                print(f"  ✓ Code compiles successfully")
            else:
                print(f"  ✗ Code compilation/execution failed")

    # Summary
    print("\n" + "=" * 70)
    print("SUMMARY")
    print("=" * 70)

    passed = sum(1 for r in results if r.success)
    total = len(results)
    mean_time = sum(r.solve_time for r in results) / total if total > 0 else 0

    print(f"\nPass Rate: {passed}/{total} ({100*passed//total}%)")
    print(f"Mean Solve Time: {mean_time:.2f}s")
    print(f"\nMethod Distribution:")
    for method, count in sorted(method_counts.items(), key=lambda x: -x[1]):
        print(f"  {method:20s}: {count:3d} problems")

    print(f"\nDetailed Results:")
    for result in results:
        status = "✓" if result.success else "✗"
        print(f"  {status} {result.problem_name:20s} | {result.solve_time:7.2f}s | {result.method:20s}")

    # Failures
    failures = [r for r in results if not r.success]
    if failures:
        print(f"\nFailures ({len(failures)}):")
        for f in failures:
            print(f"  - {f.problem_name}: {f.error[:80] if f.error else 'unknown error'}")

    # JSON output for automation
    output = {
        "pass_count": passed,
        "total": total,
        "mean_solve_time": mean_time,
        "method_distribution": method_counts,
        "results": [asdict(r) for r in results],
        "any_failures": len(failures) > 0,
    }

    output_file = Path("/Users/bobbyprice/projects/nCPU/nsynth/verify_stage4_results.json")
    with open(output_file, "w") as f:
        json.dump(output, f, indent=2)

    print(f"\nResults saved to {output_file}")

    return 0 if passed == total else 1

if __name__ == "__main__":
    sys.exit(main())

File: nsynth/src/bin/verify_stage4.rs
use mog_synth::benchmark::get_benchmark;
use mog_synth::solver::solve_problem;
use std::collections::HashMap;
/// Stage 4 Time-Parameterized Synthesis Verification
///
/// End-to-end verification for representative benchmarks (fibonacci, factorial, triangular, linear, poly).
/// Reports: (pass_count / total, mean_solve_time, method_distribution, any_failures).
use std::time::Instant;

#[derive(Clone, Debug)]
struct VerificationResult {
    problem_name: String,
    success: bool,
    solve_time_secs: f64,
    method: String,
    code_length: usize,
    error: Option<String>,
}

fn verify_stage4_benchmarks() {
    // Representative Stage 4 time-parameterized benchmarks
    let test_cases = vec![
        "fibonacci_v0",
        "factorial_v0",
        "triangular_check_v0",
        "polynomial_v0",
        "collatz_steps_v0",
    ];

    let mut results = Vec::new();
    let mut method_counts: HashMap<String, usize> = HashMap::new();

    println!("============================================================");
    println!("Stage 4 Time-Parameterized Synthesis Verification");
    println!("============================================================\n");

    // Get all problems from the benchmark factory
    let all_problems = get_benchmark(1);

    // Debug: Find and print matching problems
    eprintln!("Looking for target problems...");
    for target in &[
        "fibonacci",
        "factorial",
        "triangular",
        "polynomial",
        "collatz",
    ] {
        let matches: Vec<&str> = all_problems
            .iter()
            .filter(|p| p.name.contains(target))
            .map(|p| p.name.as_str())
            .collect();
        if !matches.is_empty() {
            eprintln!("  Matching '{}': {:?}", target, matches);
        }
    }
    eprintln!("Total problems: {}\n", all_problems.len());

    for problem_name in &test_cases {
        println!("Synthesizing {}...", problem_name);

        let problem = all_problems.iter().find(|p| &p.name == *problem_name);

        let result = match problem {
            Some(prob) => {
                let start = Instant::now();
                let solve_result = solve_problem(prob);
                let elapsed = start.elapsed().as_secs_f64();

                let success = solve_result.success;
                let method = solve_result.method.clone();
                let code = solve_result.code.clone();

                {
                    let key = if success {
                        method.clone()
                    } else {
                        "failed".to_string()
                    };
                    *method_counts.entry(key).or_insert(0) += 1;
                }

                let status = if success {
                    "✓ Success"
                } else {
                    "✗ Failed "
                };
                println!(
                    "  {} | Time: {:.2}s | Method: {} | Code: {} bytes",
                    status,
                    elapsed,
                    method,
                    code.len()
                );

                if let Some(ref err) = solve_result.error {
                    println!("    Error: {}", err);
                }

                VerificationResult {
                    problem_name: problem_name.to_string(),
                    success,
                    solve_time_secs: elapsed,
                    method,
                    code_length: code.len(),
                    error: solve_result.error,
                }
            }
            None => {
                method_counts
                    .entry("not_found".to_string())
                    .and_modify(|c| *c += 1)
                    .or_insert(1);
                println!("  ✗ Not Found in benchmark set");

                VerificationResult {
                    problem_name: problem_name.to_string(),
                    success: false,
                    solve_time_secs: 0.0,
                    method: "not_found".to_string(),
                    code_length: 0,
                    error: Some(format!(
                        "Problem '{}' not found in benchmark set",
                        problem_name
                    )),
                }
            }
        };

        results.push(result);
    }

    // Summary
    let passed = results.iter().filter(|r| r.success).count();
    let total = results.len();
    let mean_time = if total > 0 {
        results.iter().map(|r| r.solve_time_secs).sum::<f64>() / total as f64
    } else {
        0.0
    };

    println!("\n============================================================");
    println!("SUMMARY");
    println!("============================================================");
    println!("\nResults:");
    println!(
        "  Pass Rate:       {}/{} ({:3}%)",
        passed,
        total,
        if total > 0 { 100 * passed / total } else { 0 }
    );
    println!("  Mean Solve Time: {:.2}s", mean_time);

    println!("\nMethod Distribution:");
    let mut sorted_methods: Vec<_> = method_counts.iter().collect();
    sorted_methods.sort_by(|a, b| b.1.cmp(a.1));
    for (method, count) in sorted_methods {
        println!("  {:20}: {:3} problems", method, count);
    }

    println!("\nDetailed Results:");
    for result in &results {
        let status = if result.success { "✓" } else { "✗" };
        println!(
            "  {} {:20} | {:7.2}s | {:20} | {:6} bytes",
            status, result.problem_name, result.solve_time_secs, result.method, result.code_length
        );
    }

    let failures: Vec<_> = results.iter().filter(|r| !r.success).collect();
    let has_failures = !failures.is_empty();
    if has_failures {
        println!("\nFailures ({}):", failures.len());
        for f in &failures {
            println!(
                "  - {}: {}",
                f.problem_name,
                f.error.as_deref().unwrap_or("unknown error")
            );
        }
    }

    println!("\n============================================================");
    println!("FINAL RESULT");
    println!("============================================================");
    println!("pass_count / total:        {}/{}", passed, total);
    println!("mean_solve_time:           {:.2}s", mean_time);
    println!("any_failures:              {}", has_failures);
    println!("method_distribution:");
    for (method, count) in method_counts.iter() {
        println!("  {:20}: {}", method, count);
    }

    std::process::exit(if has_failures { 1 } else { 0 });
}

fn main() {
    verify_stage4_benchmarks();
}
